mixedversion: introduce mutator interface and internal API

renatolabs opened the pull request:
This commit introduces the internal API to be used by future
implementations of the `mutator` interface. Mutators are components
that change the upgrade test plan generated by the planner in order to
introduce *optional* behaviours in a test. This could mean moving a
certain step around in the plan (if it is safe to do) or, most
commonly, inserting new steps that perform an action that should not
affect the outcome of the test.

The main new data structure added in this commit is the `stepSelector`
struct, providing an internal API that mutators will be able to use in
order to find parts of the test plan they are interested in and being
able to apply mutations to it. In a future commit, these mutations
will be applied to the original test plan.

Epic: none

Release note: None

**mixedversion: apply mutations generated by `mutator` implementations**
This commit introduces a list of `planMutators` (currently empty) and
adds the logic in the test planner to iterate over these mutators,
probabilistically enable *some* of them, and then apply the mutations
they generate on the original test plan.

Epic: none

Release note: None

Commit:
mixedversion: introduce `mutator` interface and internal API

This commit introduces the internal API to be used by future
implementations of the `mutator` interface. Mutators are components
that change the upgrade test plan generated by the planner in order to
introduce *optional* behaviours in a test. This could mean moving a
certain step around in the plan (if it is safe to do) or, most
commonly, inserting new steps that perform an action that should not
affect the outcome of the test.

The main new data structure added in this commit is the `stepSelector`
struct, providing an internal API that mutators will be able to use in
order to find parts of the test plan they are interested in and being
able to apply mutations to it. In a future commit, these mutations
will be applied to the original test plan.

Epic: none

Release note: None

Commit:
mixedversion: apply mutations generated by `mutator` implementations

This commit introduces a list of `planMutators` (currently empty) and
adds the logic in the test planner to iterate over these mutators,
probabilistically enable *some* of them, and then apply the mutations
they generate on the original test plan.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/117993)
<!-- Reviewable:end -->

renatolabs commented:
### Context: mixedversion test mutators

This PR is part of a larger effort to introduce the concept of plan _mutators_ to the mixedversion framework. For initial discussion, see [internal thread](https://cockroachlabs.slack.com/archives/C01UDP33CCS/p1700252725103829).

In short, mutators are components that modify (i.e., mutate) the original test plan generated by the mixedversion planner. These changes are supposed to introduce new test scenarios in ways that do not interfere with the vast majority of tests. Tests will have the ability to individually enable or disable specific mutators as necessary; by default, all mutators will apply under some probability.

There is a working implementation of the entire idea in the [rc/mixedversion-plan-mutators](https://github.com/renatolabs/cockroach/tree/rc/mixedversion-plan-mutators) branch, in case it is helpful to see parts of the entire picture. That branch includes the proposed mutator API and an implementation of a mutator that, when merged, will close #111595.

#### History

* #115118: where we introduced the `upgradePlan` data structure and allows the upgrade to isolate the steps associated with each upgrade performed during the test.
* #117498: where we removed the `ID()` function from `singleStepProtocol`, allowing IDs to be assigned to steps later in the process, after the entire plan has been generated (and, now, mutations have been applied).

RaduBerinde reviewed (approved):
I don't have all the context but looks very good to me overall!

I posted various nits, feel free to incorporate or ignore.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117993#-:-Np12Vis-Xv3-I5Tg8sM:buimb75)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, and @renatolabs)

___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go` line 110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/117993#-Np0zfhi1FNDKoE-tIME:-Np0zfhi1FNDKoE-tIMF:bzgzfbt) ([raw file](https://github.com/cockroachdb/cockroach/blob/37ae261afcd83896b400fc4453b2eabb8ac314a1/pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go#L110)):*
> ```Go
> 	// includes its implementation.
> 	mutation struct {
> 		reference *singleStep
> ```

Should this just be an ID? Seems strange that we'd need the full impl if it's just a reference step
___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go` line 118 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/117993#-Np0zxM87VlfkVlbTZrx:-Np0zxM87VlfkVlbTZry:bn1y7ha) ([raw file](https://github.com/cockroachdb/cockroach/blob/37ae261afcd83896b400fc4453b2eabb8ac314a1/pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go#L118)):*
> ```Go
> 	// implementations to select the steps in the test plan that they
> 	// wish to mutate.
> 	stepSelector struct {
> ```

Why isn't this just `type stepSelector []*singleStep`?

Also, nit but I'd define it right above its methods
___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go` line 147 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/117993#-Np0yrmE8_YoJEgvj3j_:-Np0yrmE8_YoJEgvj3ja:b-nsngab) ([raw file](https://github.com/cockroachdb/cockroach/blob/37ae261afcd83896b400fc4453b2eabb8ac314a1/pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go#L147)):*
> ```Go
> 	AfterUpgradeFinalizedStage
> 
> 	mutationInsertBefore = mutationOp(iota)
> ```

[nit] these should be defined right after the `mutationOp` for visibility. Separating into large "type/const" blocks is counter-productive IMO. Same for the UpgradeStage above

Also, the typical syntax is `mutationInsertBefore mutationOp = iota`
___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go` line 136 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/117993#-Np1-hkeAdWBdugGnGYT:-Np1-hkeAdWBdugGnGYU:b39rn1k) ([raw file](https://github.com/cockroachdb/cockroach/blob/f26d589d35b9dc0da3d816a162a502c16ad61dc6/pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go#L136)):*
> ```Go
> 	// along with some step metadata, providing functionality to
> 	// determine the test context when inserting new steps in a plan.
> 	stepIndex struct {
> ```

[nit] `stepIndex []singleStepInfo`
___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go` line 470 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/117993#-Np1-tt7-lIOH0Uy3Avs:-Np1-tt7-lIOH0Uy3Avt:b-dkicxj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f26d589d35b9dc0da3d816a162a502c16ad61dc6/pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go#L470)):*
> ```Go
> // the given function `f` for every `singleStep` (i.e., every step
> // that actually performs an action). The function should return a
> // list of testSteps that are be used in the plan instead of the given
> ```

"that replace the given step in the plan"
___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go` line 472 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/117993#-Np1-yinBSTma-XtGEdB:-Np1-yinBSTma-XtGEdC:b1qexoc) ([raw file](https://github.com/cockroachdb/cockroach/blob/f26d589d35b9dc0da3d816a162a502c16ad61dc6/pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go#L472)):*
> ```Go
> // list of testSteps that are be used in the plan instead of the given
> // step.
> func (plan *TestPlan) mapSingleStep(f func(*singleStep, bool) []testStep) {
> ```

[nit] `mapSingleSteps`?
___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go` line 503 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/117993#-Np10dny7giI367W2Q4f:-Np10dnzCkDiDXk9-os0:b-3rbwvd) ([raw file](https://github.com/cockroachdb/cockroach/blob/f26d589d35b9dc0da3d816a162a502c16ad61dc6/pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go#L503)):*
> ```Go
> 			// have a "concurrent run" of a single step, so this
> 			// simplification makes the test plan more understandable.
> 			if s.label == genericLabel && len(newSteps) == 1 {
> ```

Should we also handle `len(newSteps) == 0`?
___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go` line 584 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/117993#-Np114fOEo5bVXoo1d9h:-Np114fOEo5bVXoo1d9i:b5piodw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f26d589d35b9dc0da3d816a162a502c16ad61dc6/pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go#L584)):*
> ```Go
> 
> 			idx := j // inserting before `step`
> 			if op == mutationInsertAfter {
> ```

[nit] consider making this a switch with a `default: panic` to help catch bugs
___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/planner_test.go` line 358 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/117993#-Np1-Hnp3vGTLvir3pM4:-Np1-Hnp3vGTLvir3pM5:b-z78str) ([raw file](https://github.com/cockroachdb/cockroach/blob/37ae261afcd83896b400fc4453b2eabb8ac314a1/pkg/cmd/roachtest/roachtestutil/mixedversion/planner_test.go#L358)):*
> ```Go
> }
> 
> func Test_stepSelectorFilter(t *testing.T) {
> ```

[nit] What's with the weird name? Why not `TestStepSelectorFilter`?


<!-- Sent from Reviewable.io -->

herkolategan reviewed (approved):
Excellent, I'm looking forward to mutators adding more coverage!

herkolategan commented on pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go:
```diff
@@ -75,6 +75,49 @@ type (
+		Generate(*rand.Rand, *TestPlan) []mutation
+	}
+
+	// mutationOp encodes the operation to be performed by a mutation.
```
Nit: This comment misled me a bit to believe `mutationOp` is the operation itself rather that how it places the mutation operation in relation to a step.

herkolategan commented on pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go:
```diff
@@ -100,6 +143,11 @@ const (
 	LastUpgradeStage
 	RunningUpgradeMigrationsStage
 	AfterUpgradeFinalizedStage
+
```
Super Nit: small part of me still like these to be in separate const blocks. But a return is probably good enough separation.

herkolategan commented on pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go:
```diff
@@ -383,50 +431,171 @@ func (up *upgradePlan) Add(steps []testStep) {
 }
 
+// newStepSelector creates a `stepSelector` instance that can be used
+// to by mutators to find a specific step or set of steps. The
```
Nit: "can be used to by mutators" s/to//

herkolategan reviewed (approved):
Wrong button 🤦‍♂️, added a few nits/comments missing in the previous review.

renatolabs reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117993)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DarrylWong, @herkolategan, and @RaduBerinde)

___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go` line 101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/117993#gh-1467589955:-Np5j8CF7HpcHw-JHwt_:b-ezikv) ([raw file](https://github.com/cockroachdb/cockroach/blob/37ae261afcd83896b400fc4453b2eabb8ac314a1/pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go#L101)):*
<details><summary><i>Previously, herkolategan (Herko Lategan) wrote…</i></summary><blockquote>

Nit: This comment misled me a bit to believe `mutationOp` is the operation itself rather that how it places the mutation operation in relation to a step.
</blockquote></details>

Updated the comment to reflect that this relates to how the change is applied, hopefully it's clearer now.
___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go` line 110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/117993#-Np0zfhi1FNDKoE-tIME:-Np5jHt81gC8J1M_SoY_:b-3e8zd7) ([raw file](https://github.com/cockroachdb/cockroach/blob/37ae261afcd83896b400fc4453b2eabb8ac314a1/pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go#L110)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Should this just be an ID? Seems strange that we'd need the full impl if it's just a reference step
</blockquote></details>

It could if these IDs were some kind of universal ID. But, in this package, they are meant more as a sequence number of steps -- see following example, where I copied part of a test plan  (IDs are in parathesis):

```
├── run "setup tpcc" (3)
├── run "run TPCC background workload" (4)
├── upgrade cluster from "v22.2.17" to "v23.1.14"
│   ├── prevent auto-upgrades by setting `preserve_downgrade_option` (5)
│   ├── upgrade nodes :1-40 from "v22.2.17" to "v23.1.14"
│   │   ├── restart node 15 with binary version v23.1.14 (6)
│   │   ├── restart node 7 with binary version v23.1.14 (7)
│   │   ├── restart node 3 with binary version v23.1.14 (8)
```

If we assigned IDs before we applied mutations (to allow mutations to reference steps by ID), we would then need to "refresh" the IDs once the mutations were applied, making it more complicated IMO.
___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go` line 118 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/117993#-Np0zxM87VlfkVlbTZrx:-Np5keZZ9IjGCgGBF8Qw:b5zopcy) ([raw file](https://github.com/cockroachdb/cockroach/blob/37ae261afcd83896b400fc4453b2eabb8ac314a1/pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go#L118)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Why isn't this just `type stepSelector []*singleStep`?

Also, nit but I'd define it right above its methods
</blockquote></details>

Heh, I guess I have this tendency to make it a struct in case we need to add more state in the future. But you're right, it gets a little simpler this way. I updated it to be just the slice.

Re: defining above methods: I think there's a personal choice element to this. I see the point of having types and functions close, but I also personally like seeing the data structures in a single place to get a quicker understanding of how data is shaped in a package.

We might decide to change this in the future, but it would probably be better to do this for all types in a separate commit.
___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go` line 146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/117993#gh-1467590640:-Np5lG6i4eIlNPHvSSrh:bbrx2dp) ([raw file](https://github.com/cockroachdb/cockroach/blob/37ae261afcd83896b400fc4453b2eabb8ac314a1/pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go#L146)):*
<details><summary><i>Previously, herkolategan (Herko Lategan) wrote…</i></summary><blockquote>

Super Nit: small part of me still like these to be in separate const blocks. But a return is probably good enough separation.
</blockquote></details>

I'll defer this to possible future work. It's in the same vein as moving type definitions as Radu suggested.
___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go` line 147 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/117993#-Np0yrmE8_YoJEgvj3j_:-Np5lU3u3nc0seAfIHav:bz6l9kt) ([raw file](https://github.com/cockroachdb/cockroach/blob/37ae261afcd83896b400fc4453b2eabb8ac314a1/pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go#L147)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] these should be defined right after the `mutationOp` for visibility. Separating into large "type/const" blocks is counter-productive IMO. Same for the UpgradeStage above

Also, the typical syntax is `mutationInsertBefore mutationOp = iota`
</blockquote></details>

Ah didn't know that syntax was more common, updated.
___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go` line 467 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/117993#gh-1467590693:-Np5nkd53BFqb02gPKWl:bj3eoea) ([raw file](https://github.com/cockroachdb/cockroach/blob/37ae261afcd83896b400fc4453b2eabb8ac314a1/pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go#L467)):*
<details><summary><i>Previously, herkolategan (Herko Lategan) wrote…</i></summary><blockquote>

Nit: "can be used to by mutators" s/to//
</blockquote></details>

Fixed, thanks.
___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go` line 136 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/117993#-Np1-hkeAdWBdugGnGYT:-Np5l7QNBxxKsFnr2weO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f26d589d35b9dc0da3d816a162a502c16ad61dc6/pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go#L136)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] `stepIndex []singleStepInfo`
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go` line 470 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/117993#-Np1-tt7-lIOH0Uy3Avs:-Np5lXE-CVqQZ7xEXxwK:b-eqkjq3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f26d589d35b9dc0da3d816a162a502c16ad61dc6/pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go#L470)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

"that replace the given step in the plan"
</blockquote></details>

Much simpler way to phrase it, thanks.
___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go` line 472 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/117993#-Np1-yinBSTma-XtGEdB:-Np5lkmC6aMRyu4lMRwM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f26d589d35b9dc0da3d816a162a502c16ad61dc6/pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go#L472)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] `mapSingleSteps`?
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go` line 503 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/117993#-Np10dny7giI367W2Q4f:-Np5mzPX6mx9WpJ8dNrs:b-g8vbxg) ([raw file](https://github.com/cockroachdb/cockroach/blob/f26d589d35b9dc0da3d816a162a502c16ad61dc6/pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go#L503)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Should we also handle `len(newSteps) == 0`?
</blockquote></details>

Currently, it should not be possible for `len(newSteps)` to be zero, barring an internal bug. The test planner should only be creating concurrent steps with >= 2 steps. We then apply mutations sequentially and, when applying, we handle the case of a single-step concurrent run (each mutation can remove at most one step).

So IMO it's better to not handle this case and, if it comes up, we know something is wrong. We could also add an assertion but that would be an overkill at the moment.
___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go` line 584 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/117993#-Np114fOEo5bVXoo1d9h:-Np5nj_q5jnXS7UITLBe:b-fznogt) ([raw file](https://github.com/cockroachdb/cockroach/blob/f26d589d35b9dc0da3d816a162a502c16ad61dc6/pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go#L584)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] consider making this a switch with a `default: panic` to help catch bugs
</blockquote></details>

Good idea, done.
___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/planner_test.go` line 358 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/117993#-Np1-Hnp3vGTLvir3pM4:-Np5nmVIDUfT5V8Yb8aK:btkjn50) ([raw file](https://github.com/cockroachdb/cockroach/blob/37ae261afcd83896b400fc4453b2eabb8ac314a1/pkg/cmd/roachtest/roachtestutil/mixedversion/planner_test.go#L358)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] What's with the weird name? Why not `TestStepSelectorFilter`?
</blockquote></details>

Heh, that might be my own convention for test functions that exercise private functions :sweat_smile:

I could rename them, but it's another of those things that should probably happen for all instances of this in a separate commit.


<!-- Sent from Reviewable.io -->

renatolabs commented:
Running a build on TC with mixed-version tests on this branch:

https://teamcity.cockroachdb.com/viewQueued.html?itemId=13711998&tab=queuedBuildOverviewTab.

renatolabs commented:
TFTRs!

bors r=RaduBerinde,herkolategan
